tests/devapps/Unity/Assets/Scripts/MSALScript.cs

```diff
@@ -0,0 +1,118 @@
+    private readonly string redirectUrl = "https://login.microsoftonline.com/common/oauth2/nativeclient";
+    private readonly string authority = "https://login.microsoftonline.com/common";
+    private readonly List<string> scopes = new List<string>() { "User.Read" };
+    private string _deviceCode = "-";
+
+    // Start is called before the first frame update
+    // https://docs.unity3d.com/ScriptReference/MonoBehaviour.html
+    async void Start()
+    {
+        await Login();
+
+    }
+
+    // Update is called once per frame
+    // https://docs.unity3d.com/ScriptReference/MonoBehaviour.html
+    void Update()
+    {
+        DeviceCodeTextField.text = $"Device code: {_deviceCode}";
+    }
+
+    private async Task Login()
+    {
+        IPublicClientApplication app = PublicClientApplicationBuilder.Create(clientId)
+            .WithAuthority(authority)
+            .WithRedirectUri(redirectUrl)
+            .WithLogging((level, message, pii) => {
+                Debug.Log($"MSAL [{level}] {pii} - {message}");
+            }, LogLevel.Verbose, true)
+            .Build();
+
+        await GetToken(); // Acquires token with device code
+        await GetToken(); // Acquires token silently
+
+        async Task GetToken()
+        {
+            AuthenticationResult authResult = null;
+
+            try
+            {
+                try
+                {
+                    var accounts = await app.GetAccountsAsync();
+                    Log("MSAL acquiring token silently.");
+                    authResult = await app.AcquireTokenSilent(scopes, accounts.FirstOrDefault()).ExecuteAsync(CancellationToken.None);
+                    Log("MSAL acquired token silently.");
+                }
+                catch (MsalUiRequiredException)
+                {
+                    Log("MSAL acquiring token with device code.");
+                    authResult = await app
+                        .AcquireTokenWithDeviceCode(
+                            scopes,
+                            deviceCodeResult => {
+                                _deviceCode = deviceCodeResult.UserCode;
+                                //CopyToClipboard(deviceCodeResult.UserCode);
+
+                                return Task.CompletedTask;
+                            })
+                        .ExecuteAsync();
+                    Log("MSAL acquired token with device code.");
+                }
+            }
+            catch (Exception ex)
+            {
+                Log($"MSAL Exception acquiring token: {ex}");
+            }
+
+            if (authResult is null)
+            {
+                Log("MSAL auth result is null.");
+            }
+            else
+            {
+                var sb = new StringBuilder();
+                sb.AppendLine($"MSAL Username={authResult.Account.Username}");
+                sb.AppendLine($"MSAL AccessToken={authResult.AccessToken}");
+                sb.AppendLine($"MSAL IdToken={authResult.IdToken}");
```

Comment:
don't print the AT.

Reply:
I think this applies to public samples? It's helpful to print ATs to decode them later in dev apps.